# Cloze Form Question Answering Using Deep Learning

Senadeera Damith Chamalke (G1702907E)<br />
Mohiuddin Mohammed Tasnim (G1702965K)<br />
Nguyen Thanh Tung (G1702496B)<br />
Jwalapuram Prathyusha (G1802075H)<br />


This project has been hosted on GitHub at https://github.com/tasnim007/Question-Answering-CNN

## Motivation
 
Machine reading comprehension or question answering is a significant research problem in the field of natural language processing. Traditional natural language processing techniques used heavy, hand-crafted pre-processing of the context, like parsing the document and the question through grammatical and syntactical parsers, extracting verb frames or predicate-argument pairs, etc.
 
Deep learning has emerged as a promising candidate for the task of question answering mainly due to the fact that deep learning models can learn features without heavy pre-processing. Therefore we thought it would be interesting to investigate the application of different deep learning techniques for this question answering task.


## Description
 
Reading comprehension is a form of a question-answering problem that tests the ability of a machine to read and understand documents by providing a model with a document/context from which the model has to find the answer to a given query. The task which we specifically focused on is called `cloze form question answering`. The speciality of **cloze form** question answering is that the query only has a single word answer, which is present in the document/passage. 

For our task, we use the CNN dataset proposed by [Hermann et al (2015)](https://arxiv.org/abs/1506.03340). It was created by using news articles from CNN as the document/passage, and generating the query by using an abstractive summary of the passage and replacing one of the words in the query with a placeholder. The task is to predict this placeholder. Since it may be possible to solve the task by training a language model on other news articles, the creators anonymise the words (persons/organizations/locations, etc.) which are possible answer candidates by replacing them with unique, numbered placeholders, henceforth referred to as “entities”.


## Data Exploration
 
The total data set of $387,420$ stories was already divided into a training set of $380,298$ stories, a validation set of $3,924$ stories and a testing set of $3,198$ stories. For the purpose of exploring this data set we analysed how the tokens and entities (candidate answers) are distributed in the passage. Also, another thing we tried to visualize is that how many times the correct answer occurs in the documents along with the distribution of tokens and entities in questions.
 
Therefore we analysed separately the statistics for these divided training, validation and testing sets of data using the histogram distributions and central tendency measures for the above mentioned scenarios with the aid of matplotlib python library for data visualization. 

**Please refer to the separate notebook [2_Exploring_Data_Codes.ipynb] for codes and details of data exploration diagram generation.**


## Analysis of Training dataset
 
In the training data set which consisted of $380,298$ examples, there were $118,432$ unique tokens and $527$ unique entities (candidate answers).

<img src="images/vs_pie_1.png"> 
 
### The Histogram for the distribution of number of tokens in training documents:

<img src="images/vs_img_1.png"> 

According to this histogram distribution, in a majority of the documents in the training corpus it can be seen that there are around $250$ to $800$ words and the distribution is skewed leftwards.
 
In a training document it can include from a minimum number of $8$ tokens to a maximum number of $2000$ tokens.
 
Three statistic values for the central tendency of this distribution of number of tokens in training documents are given below.
 
`The mode =  474`

`The median = 700.0`

`The mean = 761.81`

### The Histogram for the distribution of number of entities in training documents.
 
<img src="images/vs_img_2.png">
 
According to this histogram distribution, in a majority of the documents in the training corpus it can be seen that there are around $30$ to $80$ entities and the distribution is skewed leftwards.
 
In a training document it can include from a minimum number of $1$ entity to a maximum number of $745$ entities.
 
Three statistic values for the central tendency of this distribution of number of entities in training documents are  given below.
 
`The mode =  34`

`The median = 49.0`

`The mean = 56.09`


### The Histogram for the distribution of number of occurrences of the answer entity in training docs.
 
<img src="images/vs_img_3.png">
 
According to this histogram distribution, in a majority of the documents in the training corpus it can be seen that the answer entity only occurs once and the distribution is skewed leftwards.
 
In a training document the answer entity can occur from a minimum of $1$ time to a maximum of $61$ times.
 
Three statistic values for the central tendency of this distribution of number of occurrences of the answer entity in training documents are  given below.
 
`The mode =  1`

`The median = 4.0`

`The mean = 6.2`

### The Histogram for the distribution of number of tokens in training questions.
 
<img src="images/vs_img_4.png">
 
According to this histogram distribution, in a majority of the questions in the training corpus it can be seen that there are around $8 $to $18$ tokens and the distribution is skewed leftwards.
 
In a training question it can include from a minimum number of $1$ tokens to a maximum number of $47$ tokens.
 
Three statistic values for the central tendency of this distribution of number of tokens in training questions are  given below.
 
`The mode =  12`

`The median = 12.0`

`The mean = 12.47`

### The histogram for the distribution of number of entities in training questions.
 
<img src="images/vs_img_5.png">
 
According to this histogram distribution, in a majority of the questions in the training corpus it can be seen that there are around $0$ to $3$ entities and the distribution is skewed leftwards.
 
In a training question it can include from a minimum number of $0$ entities to a maximum number of $9$ entities.
 
Three statistic values for the central tendency of this distribution of number of entities in training documents are  given below.
 
`The mode =  1`

`The median = 1.0`

`The mean = 1.11`




## Analysis of Validation dataset
 
In the validation data set which consisted of $3,924$ examples, there were $24,204$ unique tokens and $187$ unique entities (candidate answers).

<img src="images/vs_pie_2.png">
 
### The Histogram for the distribution of number of tokens in validation documents.

<img src="images/vs_img_6.png">

According to this histogram distribution, in a majority of the documents in the validation corpus it can be seen that there are around $300$ to $700$ words and the distribution is skewed leftwards.
 
In a validation document it can include from a minimum number of 94 tokens to a maximum number of 1989 tokens.
 
Three statistic values for the central tendency of this distribution of number of tokens in validation documents are given below.
 
`The mode =  1102`

`The median = 710.0`

`The mean = 762.74`

### The Histogram for the distribution of number of entities in validation documents.
 
<img src="images/vs_img_7.png">
 
According to this histogram distribution, in a majority of the documents in the validation corpus it can be seen that there are around $20$ to $70$ entities and the distribution is skewed leftwards.
 
In a validation document it can include from a minimum number of $2$ entity to a maximum number of $223$ entities.
 
Three statistic values for the central tendency of this distribution of number of entities in validation documents are  given below.
 
`The mode =  40`

`The median = 48.0`

`The mean = 56.57`

### The Histogram for the distribution of number of occurrences of the answer entity in validation docs.
 
<img src="images/vs_img_8.png">
 
According to this histogram distribution, in a majority of the documents in the validation corpus it can be seen that the answer entity is only occurred $1$ time and the distribution is skewed leftwards.
 
In a validation document the answer entity can occur from a minimum number of $1$ times to a maximum number of $44$ times.
 
Three statistic values for the central tendency of this distribution of number of occurrences of the answer entity in validation documents are  given below.
 
`The mode =  1`

`The median = 4.0`

`The mean = 6.480632008154944`

### The Histogram for the distribution of number of tokens in validation questions.
 
<img src="images/vs_img_9.png">
 
According to this histogram distribution, in a majority of the questions in the validation corpus it can be seen that there are around $10$ to $20$ tokens and the distribution is skewed leftwards.
 
In a validation question it can include from a minimum number of $3$ tokens to a maximum number of $37$ tokens.
 
Three statistic values for the central tendency of this distribution of number of tokens in validation questions are  given below.
 
`The mode =  12`

`The median = 13.0`

`The mean = 13.289755351681958`

### The Histogram for the distribution of number of entities in validation questions.
 
<img src="images/vs_img_10.png">
 
According to this histogram distribution, in a majority of the questions in the validation corpus it can be seen that there are around $0$ to $3$ entities and the distribution is skewed leftwards.
 
In a validation question it can include from a minimum number of $0$ entities to a maximum number of $8$ entities.
 
Three statistic values for the central tendency of this distribution of number of entities in validation documents are  given below.
 
`The mode =  1`

`The median = 1.0`

`The mean = 1.2313965341488278`

 

 




## Analysis of Testing dataset
 
In the testing data set which consisted of $3,198$ examples, there were $23,145$ unique tokens and $396$ unique entities (candidate answers).

<img src="images/vs_pie_3.png">
 
### The Histogram for the distribution of number of tokens in testing documents.
 
<img src="images/vs_img_11.png">
 
According to this histogram distribution, in a majority of the documents in the testing corpus it can be seen that there are around $250$ to $800$ words and the distribution is skewed leftwards.
 
In a testing document it can include from a minimum number of $89$ tokens to a maximum number of $1989$ tokens.
 
Three statistic values for the central tendency of this distribution of number of tokens in testing documents are given below.
 
`The mode =  1095`

`The median = 652.0`

`The mean = 716.442464040025`
 
### The Histogram for the distribution of number of entities in testing documents.
 
<img src="images/vs_img_12.png">
 
According to this histogram distribution, in a majority of the documents in the testing corpus it can be seen that there are around $1$ to $80$ entities and the distribution is skewed leftwards.
 
In a testing document it can include from a minimum number of $4$ entity to a maximum number of $584$ entities.
 
Three statistic values for the central tendency of this distribution of number of entities in testing documents are  given below.
 
`The mode =  31`

`The median = 43.0`

`The mean = 51.515322076297686`
 
### The Histogram for the distribution of number of occurrences of the answer entity in testing docs.
 
<img src="images/vs_img_13.png">
 
According to this histogram distribution, in a majority of the documents in the testing corpus it can be seen that the answer entity is only occurred $1$ time and the distribution is skewed leftwards.
 
In a testing document the answer entity can occur from a minimum number of $1$ times to a maximum number of $47$ times.
 
Three statistic values for the central tendency of this distribution of number of occurrences of the answer entity in testing documents are  given below.
 
`The mode =  1`

`The median = 4.0`

`The mean = 6.363352095059412`
 
### The Histogram for the distribution of number of tokens in testing questions.
 
<img src="images/vs_img_14.png">
 
According to this histogram distribution, in a majority of the questions in the testing corpus it can be seen that there are around $8$ to $15$ tokens and the distribution is skewed leftwards.
 
In a testing question it can include from a minimum number of $3$ tokens to a maximum number of $37$ tokens.
 
Three statistic values for the central tendency of this distribution of number of tokens in testing questions are  given below.
 
`The mode =  13`

`The median = 13.0`

`The mean = 13.72514071294559`
 
### The Histogram for the distribution of number of entities in testing questions.
 
<img src="images/vs_img_15.png">
 
According to this histogram distribution, in a majority of the questions in the testing corpus it can be seen that there are around $0$ to $2$ entities and the distribution is skewed leftwards.
 
In a testing question it can include from a minimum number of $0$ entities to a maximum number of $6$ entities.
 
Three statistic values for the central tendency of this distribution of number of entities in testing documents are  given below.
 
`The mode =  1`

`The median = 1.0`

`The mean = 1.1106941838649156`
 


One observation in the whole data set is that most frequent words are the articles like “the”, “to”. “of”, “an”, “a” and “in” as expected.
 
Comparing all these distributions with respect to tokens and entities in all training, validation, and testing data sets we can see that there is a common left skewness and in general, all the distributions tend to be similar in the 3 sets of training, validation and testing data. Therefore, we can assume that the training of our models will work for the validation and testing sets. Also, another observation is that in all the 3 sets most of the times the answer entity only occurs one time in the documents.

# Attention Based Models Code and Description

We hope you are running on the virtual environment mentioned in the course [Github repository](https://github.com/xbresson/CE7454_2018). If not please run the `environment.yml` file. 


If you are running for the first time please pip install the `tarfile` by executing the following command. It will be needed for extracting the downloaded `CNN dataset`.

In [ ]:
!pip install tarfile

### Import Required Libraries:
First we need to import all the required libraries.

In [1]:
import numpy as np
import glob
import os
import sys
import time
import datetime
import io

import torch
from torch.autograd import Variable
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence
from torch.nn.utils import clip_grad_norm_

import zipfile
import tarfile
import gzip
import logging
from collections import Counter

from sklearn import metrics

### Use GPU:

It is highly recommended to run this code on GPU.

- For *Unidirectional Attention Model* time for a single epoch on GPU (Nvidia GTX 1080 Ti): 0.7hr
- For *Bidirectional Attention Model* time for a single epoch on GPU (Nvidia GTX 1080 Ti): 10.2hrs

In [2]:
gpu_id = 0 #select gpu

os.environ["CUDA_DEVICE_ORDER"] = "PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"] = str(gpu_id)  

if torch.cuda.is_available():
    print('cuda available')
    dtypeFloat = torch.cuda.FloatTensor
    dtypeLong = torch.cuda.LongTensor
    
else:
    print('cuda not available')
    gpu_id = -1
    dtypeFloat = torch.FloatTensor
    dtypeLong = torch.LongTensor
    
    

cuda available


### Download the CNN Dataset:

Our CNN dataset's default directory is `./data/cnn/`. There are three files: `train.txt`, `test.txt`, and `dev.txt`.
The following function checks whether the CNN dataset is in the default directory or not. If not, then it will download the dataset from [here](http://cs.stanford.edu/~danqi/data/cnn.tar.gz) and extract it to the default directory.

In [ ]:
def check_CNN_dataset_exists(path_data='./data/'):
    flag_train_data = os.path.isfile(path_data + 'cnn/train.txt')  
    flag_test_data = os.path.isfile(path_data + 'cnn/test.txt') 
    flag_dev_data = os.path.isfile(path_data + 'cnn/dev.txt') 
    if flag_train_data==False or flag_test_data==False or flag_dev_data==False:
        print('CNN dataset missing - downloading...')
        if not os.path.exists(path_data):
            os.makedirs(path_data)
        url = "http://cs.stanford.edu/~danqi/data/cnn.tar.gz"
        os.system('wget http://cs.stanford.edu/~danqi/data/cnn.tar.gz -P ./data/')
        tar = tarfile.open('./data/cnn.tar.gz', "r:gz")
        tar.extractall('./data/')
        tar.close()
    else:
        print("CNN dataset is already there!")
        

check_CNN_dataset_exists()

### Load the train, test, and dev data:

Original entity markers' indices are arbitrarily generated. [Danqi et. al.](https://arxiv.org/pdf/1606.02858v2.pdf) reported that relabeling the entity markers based on their first occurrence in the passage and question makes training to converge faster.

Each of the `train.txt`, `test.txt`, and `dev.txt` file contains contents in the format: `Question, Answer, Document`.

In the `load_data()` function, we read a file (`train.txt/test.txt/dev.txt`) and return the all the documents, questions, and answers in separate lists. If `relabeling=True`, we relabel the entity markers based on their first occurrence as stated above.


In [3]:
def load_data(in_file, max_example=None, relabeling=True):
    """
        load CNN data from {train | dev | test}.txt
        max_example: if it is None, all the examples will be read.
        relabeling: relabel the entities by their first occurence if it is True.
    """

    documents = []
    questions = []
    answers = []
    num_examples = 0
    f = open(in_file, 'r', encoding='utf-8')
    while True:
        line = f.readline()
        if not line:
            break
        question = line.strip().lower()
        answer = f.readline().strip()
        document = f.readline().strip().lower()

        if relabeling:
            q_words = question.split(' ')
            d_words = document.split(' ')
            assert answer in d_words

            entity_dict = {}
            entity_id = 0
            for word in d_words + q_words:
                if (word.startswith('@entity')) and (word not in entity_dict):
                    entity_dict[word] = '@entity' + str(entity_id)
                    entity_id += 1

            q_words = [entity_dict[w] if w in entity_dict else w for w in q_words]
            d_words = [entity_dict[w] if w in entity_dict else w for w in d_words]
            answer = entity_dict[answer]

            question = ' '.join(q_words)
            document = ' '.join(d_words)

        questions.append(question)
        answers.append(answer)
        documents.append(document)
        num_examples += 1

        f.readline()
        if (max_example is not None) and (num_examples >= max_example):
            break
    f.close()
    
    print('#Examples: %d' % len(documents))
    return (documents, questions, answers)

### Build the Dictionary

In the `build_dict()` function, we build a dictionary of vocabularies. For building the dictionary, we keep the most frequent 120K words. We tried with smaller vocabulary size like 50K, but it did not give the optimal result. 

In [4]:
def build_dict(sentences, max_words=120000):
    """
        Build a dictionary for the words in `sentences`.
        Only the max_words ones are kept and the remaining will be mapped to <UNK>.
    """
    word_count = Counter()
    for sent in sentences:
        for w in sent.split(' '):
            word_count[w] += 1

    ls = word_count.most_common(max_words)
    print('#Words: %d -> %d' % (len(word_count), len(ls)))
    for key in ls[:5]:
        print(key)
    print('...')
    for key in ls[-5:]:
        print(key)

    # leave 0 to UNK
    # leave 1 to delimiter |||
    
    vocab_dict = {w[0]: index + 2 for (index, w) in enumerate(ls)}
    vocab_dict['<UNK>'] = 0
    vocab_dict['<PAD>'] = 1
    
    
    return vocab_dict

### Vectorize the Data (train/test/dev):

We need to vectorize our data (documents/questions/answers). For documents and questions, we use the `word_dict` dictionary for vectorizing, while for vectorizing answers we use the `entity_dict` dictionary. Basically, the indices of the dictionaries corresponding to the relevant words are replaced in input data in order to process the inputs into our deep learning models. Meanwhile, the maximum document and question lengths are also calculated.

In [5]:
def vectorize(examples, word_dict, entity_dict):
    """
        Vectorize `examples`.
        in_x1, in_x2: sequences for document and question respecitvely.
        in_y: label
        in_l: whether the entity label occurs in the document.
    
    """
    in_x1 = []
    in_x2 = []
    in_l = np.zeros((len(examples[0]), len(entity_dict)))#.astype(config._floatX)
    in_y = []
    
    max_d = 0
    max_q = 0
    
    
    for idx, (d, q, a) in enumerate(zip(examples[0], examples[1], examples[2])):
        d_words = d.split(' ')
        q_words = q.split(' ')
        assert (a in d_words)
        seq1 = [word_dict[w] if w in word_dict else 0 for w in d_words]
        seq2 = [word_dict[w] if w in word_dict else 0 for w in q_words]
        
        if max_d < len(seq1):
            max_d = len(seq1)
        if max_q < len(seq2):
            max_q = len(seq2)
        
        if (len(seq1) > 0) and (len(seq2) > 0):
            in_x1.append(seq1)
            in_x2.append(seq2)
            in_l[idx, [entity_dict[w] for w in d_words if w in entity_dict]] = 1.0
            in_y.append(entity_dict[a] if a in entity_dict else 0)
        if (idx % 10000 == 0):
            print('Vectorization: processed %d / %d' % (idx, len(examples[0])))


    return in_x1, in_x2, in_l, in_y, max_d, max_q

### Create Mini Batches:

In order to get the advantage of using GPUs for running our models, we create minibatches of data before inputting them to our model. From this function, we can create the minibatches of data according to desired minibatch sizes.

In [6]:
def get_minibatches(n, minibatch_size, shuffle=False):
    idx_list = np.arange(0, n, minibatch_size)
    if shuffle:
        np.random.shuffle(idx_list)
    minibatches = []
    for idx in idx_list:
        minibatches.append(np.arange(idx, min(idx + minibatch_size, n)))
    return minibatches

### Prepare Data with Padding:

This function is used to pad the data inputs which are less lengthier than the maximum sequence length.

In [7]:
def prepare_data(seqs, max_l):
    lengths = [len(seq) for seq in seqs]
    n_samples = len(seqs)
    #max_len = np.max(lengths)
    x = np.zeros((n_samples, max_l)).astype('int32')
    #x_len = np.zeros((n_samples, 1)).astype(config._floatX)
    for idx, seq in enumerate(seqs):
        x[idx, :lengths[idx]] = seq
        #x_len[idx,0] = len(seq)
    return x, np.array(lengths)


### Genrating Final Input  Data to Model:

This function calls the function to prepare minibatches and then the padding function to pad the the input data accordingly inorder to prepare the final input data sets.

In [8]:
def gen_examples(x1, x2, l, y, batch_size, max_d, max_q):
    """
        Divide examples into batches of size `batch_size`.
    """
    minibatches = get_minibatches(len(x1), batch_size)
    all_ex = []
    for minibatch in minibatches:
        mb_x1 = [x1[t] for t in minibatch]
        mb_x2 = [x2[t] for t in minibatch]
        #mb_l = l[minibatch]
        mb_y = [y[t] for t in minibatch]
        mb_y = np.array(mb_y)
        mb_x1, x1_len = prepare_data(mb_x1, max_d)
        mb_x2, x2_len = prepare_data(mb_x2, max_q)
        all_ex.append((mb_x1, x1_len, mb_x2, x2_len, mb_y))
    return all_ex

### Generate Pretrained Embeddings:

We tried with three types of pretrained embeddings: [**Glove** from Stanford University](https://nlp.stanford.edu/projects/glove/), [**Fasttext** from facebook](https://fasttext.cc/), and `uniform random embeddings`. We used embedding dimension of 300. Following functions generates pretrained embeddings from glove/fasttext/random according to our specification.

The default directory for glove pretrained embeddings is `./data/glove/`.
The default directory for glove pretrained embeddings is `./data/fasttext/`
If you want to use pretrained embeddings (glove/fasttext) and they are not in the default directory, it will be downloaded by the following codes. 

**Caution:<font color='red'> Dowloading pretrained embeddings may take some times. </font>**




In [9]:
def gen_embeddings_glove(word_dict, dim, in_file=None):
    """
        Generate an initial embedding matrix for `word_dict` from glove pretrained embeddings.
        If an embedding file is not given or a word is not in the embedding file,
        a randomly initialized vector will be used.
    """
    if os.path.isfile(in_file)==False:
        print('Glove pretrained embedding missing.. Downloading...')
        if not os.path.exists('./data/glove/'):
            os.makedirs('./data/glove/')
        os.system('wget http://nlp.stanford.edu/data/glove.6B.zip -P ./data/glove/')
        zip_ref = zipfile.ZipFile('./data/glove/glove.6B.zip', 'r')
        zip_ref.extractall('./data/glove/')
        zip_ref.close()
        
    
    num_words = len(word_dict) + 2
    embeddings = np.random.uniform(size=(num_words, dim))
    print('Embeddings: %d x %d' % (num_words, dim))

    if in_file is not None:
        print('Loading embedding file: %s' % in_file)
        pre_trained = 0
        for line in open(in_file, encoding='utf-8').readlines():
            sp = line.split()
            
            #print("Length = ",len(sp))
            
            assert len(sp) == dim + 1 
            if sp[0] in word_dict:
                pre_trained += 1
                embeddings[word_dict[sp[0]]] = [float(x) for x in sp[1:]]
        print('Pre-trained: %d (%.2f%%)' %
              (pre_trained, pre_trained * 100.0 / num_words))
    return embeddings

def gen_embeddings_fasttext(word_dict, dim, in_file=None):
    """
        Generate an initial embedding matrix for `word_dict`from fasttext pretrained embeddings.
        If an embedding file is not given or a word is not in the embedding file,
        a randomly initialized vector will be used.
    """
    if os.path.isfile(in_file)==False:
        print('Fasttext pretrained embedding missing.. Downloading...')
        if not os.path.exists('./data/fasttext/'):
            os.makedirs('./data/fasttext/')
        os.system('wget https://s3-us-west-1.amazonaws.com/fasttext-vectors/wiki.en.vec -P ./data/fasttext/')

    num_words = len(word_dict) + 2
    embeddings = np.random.uniform(size=(num_words, dim))
    print('Embeddings: %d x %d' % (num_words, dim))

    if in_file is not None:
        print('Loading fasttext embedding file: %s' % in_file)
        pre_trained = 0
        with io.open(in_file, 'r', encoding='utf-8', newline='\n', errors='ignore') as f:
            for i, line in enumerate(f):
                if i == 0:
                    split = line.split()
                    assert len(split) == 2
                    #assert _emb_dim_file == int(split[1])
                else:
                    word, vect = line.rstrip().split(' ', 1)        
                    if word in word_dict:
                        pre_trained += 1
                        vect = np.fromstring(vect, sep=' ')
                        #print(vect.shape)
                        embeddings[word_dict[word]] = vect
        print('Pre-trained: %d (%.2f%%)' % (pre_trained, pre_trained * 100.0 / num_words))
    return embeddings

def gen_embeddings_random(word_dict, dim):
    """
        Generate an initial embedding matrix for `word_dict` using uniform random.
    """
    num_words = len(word_dict) + 2
    embeddings = np.random.uniform(size=(num_words, dim))
    print('Embeddings: %d x %d' % (num_words, dim))
    return embeddings

def gen_embeddings(word_dict, dim, in_file=None, init_from='random'):
    """
        Generate an initial embedding matrix for `word_dict`
    """
    if init_from == 'glove':
        return gen_embeddings_glove(word_dict, dim, in_file)
    elif init_from == 'fasttext':
        return gen_embeddings_fasttext(word_dict, dim, in_file)
    else:
        return gen_embeddings_random(word_dict, dim)

# Unidirectional Attention Model

The following figure depicts our `Unidirectional Attention Model`.

<img src="images/unidirectional_attention.jpg"> 
Figure: Unidirectional Attention Model

We first convert each word of the passage to the corresponding **word embedding vectors**. We run our model with both random and pretrained word embeddings like `glove` and `fasttext`. Our word embeddings are trainable in both of the cases. 

The word embeddings do not capture the context of the words in the sentence. For this reason, we use Recurrent Neural Networks (RNN). We use bidirectional RNNs. The benefit of bidirectional RNN over unidirectional one is that *in bidirectional RNNs a word does not only have the previous words information from the forward RNNs but also have the information from the future words from the backward RNNs*. By concatenating the outputs of the forward and the backward RNNs, we get the **Contextual Embeddings** for each word.

For the RNNs, we tried with both LSTMs and GRUs.

Similarly for questions, we first convert them to word embeddings. Then by using bidirectional RNNs, we create the **Query Vector** by taking only the last RNN output. Since our query size is small (One single sentence), the bidirectional RNNs (LSTM/GRU) can retain all the information into a single vector. Thus the Query Vector has all the information regarding the corresponding query input. 

Now the **goal** is to *compare the Query Vector and all the contextual embeddings, and select the pieces of information that are relevant to the question*. For this purpose, we use the `attention mechanism`. We use the *Multiplicative Attention* mechanism to pay attention from the query vector to each of the passage words represented by the contextual embeddings. This gives us the **Attention Scores**. To compute the probability distribution of the attention scores, we use **softmax**. 

Now we have the **Attention Distribution** values. We compute the **Weighted Score** by *multiplying the attention score to their respective contextual embeddings and summing them up*.

Finally, we use a *linear layer* to get the **Logits**.


We use **Cross Entropy Loss** for the loss calculation. 

For optimization, we use  **Stochastic Gradient Descent (SGD)** starting with learning rate $0.1$. We reduce the learning rate by a factor of $1.2$, if `dev_loss > 0.99* dev_loss_old`.

We use minibatch SGD with batch size $128$.

In [11]:
class Unidirectional_Attention_Model(nn.Module):

    def __init__(self, embeddings, hidden_size, output_size, num_layers=1, drp_rate=0.25):
        super(Unidirectional_Attention_Model, self).__init__()
        
        self.vocab_size = embeddings.shape[0]
        self.embedding_size = embeddings.shape[1]
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.num_layers = num_layers
        self.drp = drp_rate
        
        weight = torch.Tensor(embeddings)
        self.embeddings = nn.Embedding.from_pretrained(weight, freeze=False)
            
        self.context_gru = nn.GRU(embedding_size, hidden_size, num_layers, 
                          batch_first=True, bidirectional=True, dropout=self.drp)
        
        self.ws = nn.Linear(hidden_size*2, hidden_size*2) # mult by 2 for bidirectional
        self.linear = nn.Linear(hidden_size*2, output_size)

        
    def forward(self, doc_x, ques_x, doc_seq_lengths, ques_seq_lengths):   
        '''
        doc_x: torch tensor. size - (batch_size, doc_seq_len)
        ques_x: torch tensor. size - (batch_size, ques_seq_len)
        doc_seq_lengths: 1d numpy array containing lengths of each document in doc_x
        ques_seq_lengths: 1d numpy array containing lengths of each question in ques_x
        
        '''
        
        def contextual_embedding(data, seq_lengths):
            # Sort by length (keep idx)
            seq_lengths, idx_sort = np.sort(seq_lengths)[::-1], np.argsort(-seq_lengths)
            idx_original = np.argsort(idx_sort)
            idx_sort = torch.from_numpy(idx_sort).type(dtypeLong)
            data = data.index_select(0, idx_sort)

            packed_input = pack_padded_sequence(data, seq_lengths, batch_first=True)
            packed_output, hidden = self.context_gru(packed_input)
            output, _ = pad_packed_sequence(packed_output, batch_first=True)
            #print("out: ", output.size(), " hid: ", hidden.size())

            #Un-sort by length
            idx_original = torch.from_numpy(idx_original).type(dtypeLong)
            output = output.index_select(0, idx_original)
            hidden = hidden.index_select(1, idx_original)
            
            return output, hidden

        doc_data = self.embeddings(doc_x) # doc_data shape: (batch_size, doc_seq_len, embedding_dim)
        ques_data = self.embeddings(ques_x) # ques_data shape: (batch_size, ques_seq_len, embedding_dim)
         
        ##For Documents/passages
        doc_output, doc_hidden = contextual_embedding(doc_data, doc_seq_lengths)
        ques_output, ques_hidden = contextual_embedding(ques_data, ques_seq_lengths)
        # output shape: (batch_size, seq_len, hidden_size * num_directions)
        # hidden shape: (num_layers * num_directions, batch_size, hidden_size)
        
        ques_fwd_h = ques_hidden[0:ques_hidden.size(0):2]
        ques_bwd_h = ques_hidden[1:ques_hidden.size(0):2]
        ques_hidden = torch.cat([ques_fwd_h, ques_bwd_h], dim=2)   
        #print("After hid: ", ques_hidden.size())
        
        q_ws = self.ws(ques_hidden) #q_ws shape:  torch.Size([1, bs, 256])
        #print("q_ws shape: ", q_ws.size()) 
        q_ws = q_ws.squeeze().unsqueeze(1) #q_ws shape:  torch.Size([bs, 1, 256])
        q_ws.transpose_(1,2) #q_ws shape:  torch.Size([bs, 256, 1])
        q_ws_p = torch.bmm(doc_output, q_ws).squeeze() # q_ws_p shape: torch.Size([bs, timesteps])
        #print("q_ws_p shape: ", q_ws_p.size())
        alpha = F.softmax(q_ws_p, dim=1) #alpha shape:  torch.Size([bs, 1808])
        #print("alpha shape: ", alpha.size())
        attention = torch.mul(alpha.unsqueeze(2), doc_output) #attention shape:  torch.Size([bs, 1808, 256])
        #print("attention shape: ", attention.size())
        attention = torch.sum(attention, dim=1) #After summing attention shape:  torch.Size([bs, 256])
        #print("After summing attention shape: ", attention.size())
        
        logits = self.linear(attention) #logits shape:  torch.Size([bs, numClasses])
        #print("logits shape: ", logits.size())
        
        return logits
    
    def loss(self, y, y_target):
        
        #loss = nn.MSELoss()(y,y_target)
        loss = nn.CrossEntropyLoss()(y,y_target)
        
        return loss


    def update(self, lr):
                
        update = torch.optim.SGD( self.parameters(), lr=lr )
        
        return update
    
    def update_learning_rate(self, optimizer, lr):
   
        for param_group in optimizer.param_groups:
            param_group['lr'] = lr

        return optimizer

### Function to Train One Single Epoch:

This function takes the input data and makes the input minibatches randomly shuffle first. Then once the epoch is over it updates the learning rate according to the given policy and outputs the training loss and accuracy.

In [12]:
def train_one_epoch(net,optimizer,tr_data):
    
    net.train()
    
    run_loss = 0
    run_nb_data = 0
    run_acc = 0
    
    
    np.random.shuffle(tr_data)
    
    for ids, (mb_x1, x1_len, mb_x2, x2_len, mb_y) in enumerate(tr_data):
        
        
        x1 = Variable( torch.LongTensor(mb_x1).type(dtypeLong) , requires_grad=False)
        x2 = Variable( torch.LongTensor(mb_x2).type(dtypeLong) , requires_grad=False)
        target = Variable( torch.LongTensor(mb_y).type(dtypeLong) , requires_grad=False)
        
        optimizer.zero_grad()
        
        pred = net.forward(x1,x2,x1_len,x2_len)
        
        loss = net.loss(pred,target)
        loss.backward()
        optimizer.step()
        
        run_loss += (loss.detach().item())*float(mb_x1.shape[0])
        run_nb_data += float(mb_x1.shape[0])
        
        
        _, batch_predicted_labels = torch.max(pred, dim=1) 
        
        
        acc = metrics.accuracy_score(np.array(target), batch_predicted_labels.cpu().numpy())
        run_acc += acc*float(mb_x1.shape[0])
        
        
    acc_tr = run_acc/run_nb_data
    loss_tr = run_loss/run_nb_data
    
        
        
    
        
    return loss_tr, acc_tr
        
    
    

### Function to Test the Validation and the Testing data:

In [13]:
def test_one_epoch(net,tst_data):
    
    net.eval()
    
    run_loss = 0
    run_nb_data = 0
    run_acc = 0
    
    for ids, (mb_x1, x1_len, mb_x2, x2_len, mb_y) in enumerate(tst_data):
        
        
        x1 = Variable( torch.LongTensor(mb_x1).type(dtypeLong) , requires_grad=False)
        x2 = Variable( torch.LongTensor(mb_x2).type(dtypeLong) , requires_grad=False)
        target = Variable( torch.LongTensor(mb_y).type(dtypeLong) , requires_grad=False)
        
        
        pred = net.forward(x1,x2,x1_len,x2_len)
        
        loss = net.loss(pred,target)
        
        run_loss += (loss.detach().item())*float(mb_x1.shape[0])
        run_nb_data += float(mb_x1.shape[0])
        
        _, batch_predicted_labels = torch.max(pred, dim=1) 
        
        
        acc = metrics.accuracy_score(np.array(target), batch_predicted_labels.cpu().numpy())
        run_acc += acc*float(mb_x1.shape[0])
        
        
    acc_dev = run_acc/run_nb_data
    loss_dev = run_loss/run_nb_data
    
    return loss_dev, acc_dev
    
    
    

In [14]:
#Loading data


fin_train = "./data/cnn/train.txt"
fin_dev = "./data/cnn/dev.txt"


train_exps = load_data(fin_train, relabeling=True)
dev_exps = load_data(fin_dev, relabeling=True)

#Examples: 380298
#Examples: 3924


In [15]:
#Building dictionaries


word_dict = build_dict(train_exps[0] + train_exps[1])
entity_markers = list(set([w for w in word_dict.keys() if w.startswith('@entity')] + train_exps[2]))
entity_markers = ['<unk_entity>'] + entity_markers
entity_dict = {w: index for (index, w) in enumerate(entity_markers)}

#Words: 118432 -> 118432
('the', 15383021)
(',', 13757778)
('.', 11782121)
('to', 7208903)
('"', 6967510)
...
('slingers', 1)
('multi-planet', 1)
('johnstons', 1)
('shir', 1)
('khurma', 1)


### Generate Embeddings
Here you can choose your initial embeddings. Use either `pretrained embeddings` (glove/fasttext) or `uniform random embeddings`. **Uncomment the one you want to use.**

In [16]:
#Generating Embeddings

embedding_size = 300

# embedding can be initialized from glove or fastest pretrained embeddings or from random ones.

#embeddings = gen_embeddings(word_dict, embedding_size, in_file='./data/glove/glove.6B.300d.txt', init_from='glove')
#embeddings = gen_embeddings(word_dict, embedding_size, in_file='./data/fasttext/wiki.en.vec', init_from='fasttext')
embeddings = gen_embeddings(word_dict, embedding_size, in_file=None, init_from='random')

Embeddings: 118436 x 300
Loading fasttext embedding file: wiki.en.vec
Pre-trained: 79469 (67.10%)


In [17]:
#Vectorizing and minibatch creation

trn_x1, trn_x2, trn_l, trn_y, max_d, max_q = vectorize(train_exps, word_dict, entity_dict)
train_data = gen_examples(trn_x1, trn_x2, trn_l, trn_y, 128, max_d, max_q)


dev_x1, dev_x2, dev_l, dev_y, _, _ = vectorize(dev_exps, word_dict, entity_dict)
dev_data = gen_examples(dev_x1, dev_x2, dev_l, dev_y, 128, max_d, max_q)

Vectorization: processed 0 / 380298
Vectorization: processed 10000 / 380298
Vectorization: processed 20000 / 380298
Vectorization: processed 30000 / 380298
Vectorization: processed 40000 / 380298
Vectorization: processed 50000 / 380298
Vectorization: processed 60000 / 380298
Vectorization: processed 70000 / 380298
Vectorization: processed 80000 / 380298
Vectorization: processed 90000 / 380298
Vectorization: processed 100000 / 380298
Vectorization: processed 110000 / 380298
Vectorization: processed 120000 / 380298
Vectorization: processed 130000 / 380298
Vectorization: processed 140000 / 380298
Vectorization: processed 150000 / 380298
Vectorization: processed 160000 / 380298
Vectorization: processed 170000 / 380298
Vectorization: processed 180000 / 380298
Vectorization: processed 190000 / 380298
Vectorization: processed 200000 / 380298
Vectorization: processed 210000 / 380298
Vectorization: processed 220000 / 380298
Vectorization: processed 230000 / 380298
Vectorization: processed 24000

In [18]:
(len(train_data[6][2][50]))

47

In [19]:
#Model parameters

lr = 0.1 #learning rate
decay_rate = 1.2
dev_loss_old = 1000000
epochs_no = 20
drp = 0.25 #drop out rate



In [20]:
#Building model


net = Unidirectional_Attention_Model(embeddings=embeddings, hidden_size=128, output_size=len(entity_dict),drp_rate=drp)
if torch.cuda.is_available():
    net.cuda()
opt = net.update(lr)

/home/damith/miniconda3/envs/deeplearning_project/lib/python3.6/site-packages/torch/nn/modules/rnn.py:38: UserWarning: dropout option adds dropout after all but last recurrent layer, so non-zero dropout expects num_layers greater than 1, but got dropout=0.25 and num_layers=1
  "num_layers={}".format(dropout, num_layers))


### Train the Model

In [21]:
# save results in a .txt file

time_stamp = datetime.datetime.now().strftime("%y-%m-%d--%H-%M-%S")
if not os.path.exists('./logs/'):
    os.makedirs('./logs/')
file_name = 'logs'+'/'+time_stamp + "unidr" + ".txt"
file = open(file_name,"w",1) 
file.write(time_stamp+'\n\n') 
mystr = "unidr_fastext_sgd"
file.write(mystr+'\n\n')

start = time.time()

for epoch in range(epochs_no):
    
    start_epoch = time.time()
    
    tr_loss, tr_acc = train_one_epoch(net,opt,train_data)
    
    dev_loss, dev_acc = test_one_epoch(net,dev_data)
    
    
    # update learning rate 
    if dev_loss > 0.99* dev_loss_old:
        lr /= decay_rate
    opt = net.update_learning_rate(opt, lr)
    dev_loss_old = dev_loss
    
    ep_details = ( 'Epoch {EP} [epoch length:{epoch_time:.0f}s | time from start:{fromstart:.1f}h] \t'
                   'lr={LR:.2e}\t'
                   'loss={train_loss:.3f}/{test_loss:.3f}\t'
                   'acc:{train_acc:.3f}/{test_acc:.3f} '.format(
                    EP=epoch,
                    epoch_time=time.time()-start_epoch,
                    fromstart=(time.time()-start)/3600,          
                    LR= lr,  
                    train_loss=tr_loss, test_loss=dev_loss,
                    train_acc=tr_acc, test_acc=dev_acc ) )
    
    
    
    
    file.write(ep_details+'\n')
    print(ep_details)
    
file.close()
    

Epoch 0 [epoch length:1852s | time from start:0.5h] 	lr=1.00e-01	loss=2.650/2.245	acc:0.286/0.392 
Epoch 1 [epoch length:1839s | time from start:1.0h] 	lr=1.00e-01	loss=2.259/2.001	acc:0.410/0.492 
Epoch 2 [epoch length:1840s | time from start:1.5h] 	lr=1.00e-01	loss=1.870/1.610	acc:0.521/0.579 
Epoch 3 [epoch length:1839s | time from start:2.0h] 	lr=1.00e-01	loss=1.580/1.420	acc:0.600/0.634 
Epoch 4 [epoch length:1848s | time from start:2.6h] 	lr=1.00e-01	loss=1.410/1.333	acc:0.641/0.659 
Epoch 5 [epoch length:1841s | time from start:3.1h] 	lr=1.00e-01	loss=1.303/1.298	acc:0.669/0.663 
Epoch 6 [epoch length:1836s | time from start:3.6h] 	lr=1.00e-01	loss=1.225/1.242	acc:0.688/0.682 
Epoch 7 [epoch length:1840s | time from start:4.1h] 	lr=1.00e-01	loss=1.168/1.229	acc:0.703/0.688 
Epoch 8 [epoch length:1836s | time from start:4.6h] 	lr=8.33e-02	loss=1.118/1.242	acc:0.715/0.681 
Epoch 9 [epoch length:1843s | time from start:5.1h] 	lr=8.33e-02	loss=1.058/1.199	acc:0.730/0.693 
Epoch 10 [

### Test the Model

In [22]:
fin_test = "./data/cnn/test.txt"
test_exps = load_data(fin_test, relabeling=True)

tst_x1, tst_x2, tst_l, tst_y, max_d, max_q = vectorize(test_exps, word_dict, entity_dict)
test_data = gen_examples(tst_x1, tst_x2, tst_l, tst_y, 128, max_d, max_q)

#Examples: 3198
Vectorization: processed 0 / 3198


In [23]:
tst_loss, tst_acc = test_one_epoch(net,test_data)
print("Test loss and accuracy = ",tst_loss,"\t",tst_acc)

Test loss and accuracy =  1.1368197198582113 	 0.7229518449030644


# Bidirectional Attention Model

The following figure depicts our `Bidirectional Attention Model`.

<img src="images/bidirectional_attention.jpg"> 
Figure: Bidirectional Attention Model

Here the first couple of layers are similar to `Unidirectional Attention Model`. Unlike unidirectional one, from the query we do not create a single *Query Vector*. Instead, we take all the output of the RNNs to create `Query Contextual Embeddings` just like the `Passage Contextual Embeddings`. 
Then we use the **Bidirectional Attention Mechanism**. Here every word in `Passage Contextual Embeddings` pays attention to every word in `Query Contextual Embeddings`, resulting `Passage to Query Attention Scores`. Similarly, every word in `Query Contextual Embeddings` pays attention to every word in `Passage Contextual Embeddings`, resulting `Query to Passage Attention Scores`. Finally, these attention scores and contextual embeddings are combined together. We followed the approach mentioned in [Seo et al.](https://arxiv.org/abs/1611.01603). The resultant one encodes *query aware representations* of context words.
Next, we use another bidirectional RNN to capture the interactions among passage words conditioned on the query.
Finally, we use three linear layers sequentially to get the **Logits**.

We use **Cross Entropy Loss** for the loss calculation. 

For optimization, we use  **Stochastic Gradient Descent (SGD)** starting with learning rate $0.1$. We reduce the learning rate by a factor of $1.2$, if `dev_loss > 0.99* dev_loss_old`.

We use minibatch SGD with batch size $8$.

In [ ]:
class Bidirectional_Attention_Model(nn.Module):

    def __init__(self, embeddings, hidden_size, output_size, maxlen, num_layers=1, drp_rate=0.25):
        super(Bidirectional_Attention_Model, self).__init__()
        
        self.vocab_size = embeddings.shape[0]
        self.embedding_size = embeddings.shape[1]
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.maxlen = maxlen
        self.num_layers = num_layers
        self.drp = drp_rate
        
        weight = torch.Tensor(embeddings)
        self.embeddings = nn.Embedding.from_pretrained(weight, freeze=False)
            
        self.context_gru = nn.GRU(embedding_size, hidden_size, num_layers=num_layers, 
                          batch_first=True, bidirectional=True, dropout=self.drp)
        
        self.sim_W = nn.Linear(hidden_size*6, 1)
        
        self.modeling_layer = nn.GRU(hidden_size*8, hidden_size, num_layers=2, 
                                     batch_first=True, bidirectional=True, dropout=self.drp)
        
        self.output_layer_1 = nn.Linear(hidden_size*10, 1)
        self.output_layer_2 = nn.Linear(maxlen, 1000)
        self.output_layer_3 = nn.Linear(1000, output_size)
        
        
    def forward(self, doc_x, ques_x, doc_seq_lengths, ques_seq_lengths):
        '''
        doc_x: torch tensor. size - (batch_size, doc_seq_len)
        ques_x: torch tensor. size - (batch_size, ques_seq_len)
        doc_seq_lengths: 1d numpy array containing lengths of each document in doc_x
        ques_seq_lengths: 1d numpy array containing lengths of each question in ques_x
        
        '''
        
        def contextual_embedding(data, seq_lengths):
            # Sort by length (keep idx)
            seq_lengths, idx_sort = np.sort(seq_lengths)[::-1], np.argsort(-seq_lengths)
            idx_original = np.argsort(idx_sort)
            idx_sort = torch.from_numpy(idx_sort).type(dtypeLong)
            data = data.index_select(0, idx_sort)

            packed_input = pack_padded_sequence(data, seq_lengths, batch_first=True)
            packed_output, hidden = self.context_gru(packed_input)
            output, _ = pad_packed_sequence(packed_output, batch_first=True)
            #print("out: ", output.size(), " hid: ", hidden.size())

            # Un-sort by length
            idx_original = torch.from_numpy(idx_original).type(dtypeLong)
            output = output.index_select(0, idx_original)
            hidden = hidden.index_select(1, idx_original)
            # output shape: (batch_size, seq_len, hidden_size * num_directions) --seq_len is the largest lengths in the minibatch
            # hidden shape: (num_layers * num_directions, batch_size, hidden_size)
            
            return output, hidden

        doc_data = self.embeddings(doc_x) # doc_data shape: (batch_size, doc_seq_len, embedding_dim)
        ques_data = self.embeddings(ques_x) # ques_data shape: (batch_size, ques_seq_len, embedding_dim)
         
        ## contextual embedding for documents and ques
        doc_output, doc_hidden = contextual_embedding(doc_data, doc_seq_lengths)
        ques_output, ques_hidden = contextual_embedding(ques_data, ques_seq_lengths)
        
        ## Attention Flow
        # Similarity Matrix calcuation
        doc_seq_len = doc_output.size(1) # T
        ques_seq_len = ques_output.size(1) # J
        shape = (doc_output.size(0), doc_seq_len, ques_seq_len, 2*self.hidden_size) # (N, T, J, 2d)
        #print("T: ", doc_seq_len, " J: ", ques_seq_len, " Shape: ", shape)
        doc_output_extra = doc_output.unsqueeze(2)  # (N, T, 1, 2d)
        doc_output_extra = doc_output_extra.expand(shape) # (N, T, J, 2d)
        ques_output_extra = ques_output.unsqueeze(1)  # (N, 1, J, 2d)
        ques_output_extra = ques_output_extra.expand(shape) # (N, T, J, 2d)
        elmwise_mul = torch.mul(doc_output_extra, ques_output_extra) # (N, T, J, 2d)
        #print("doc: ", doc_output_extra.size(), " ques: ", ques_output_extra.size(), " elem: ", elmwise_mul.size())
        cat_data = torch.cat((doc_output_extra, ques_output_extra, elmwise_mul), 3) # (N, T, J, 6d), [h;u;h◦u]
        similarity_matrix = self.sim_W(cat_data).squeeze() # (N, T, J)
        #print("cat: ", cat_data.size(), " similarity_matrix: ", similarity_matrix.size())
        
        
        a = F.softmax(similarity_matrix, dim=2) # (bs, T, J)
        doc2ques_attention = torch.bmm(a, ques_output) # (bs, T, 2*d)
        #print("a: ", a.size(), " doc2ques: ", doc2ques_attention.size())
        b = F.softmax(torch.max(similarity_matrix, dim=2)[0], dim=1).unsqueeze(1) # (bs, 1, T)
        ques2doc_attention = torch.bmm(b, doc_output).squeeze()  # (bs, 2d)
        ques2doc_attention = ques2doc_attention.unsqueeze(1).expand(-1, doc_seq_len, -1) # (bs, T, 2*d)
        # q2c_att = torch.stack([q2c_att] * c_len, dim=1)

        G = torch.cat((doc_output, doc2ques_attention, doc_output.mul(doc2ques_attention), 
                       doc_output.mul(ques2doc_attention)), 2) # (bs, T, 8d)
        #print("b: ", b.size(), " ques2doc: ", ques2doc_attention.size(), " G: ", G.size())
        
        ## Modeling Layer
        M, _h = self.modeling_layer(G) # M: (bs, T, 2d)
        
        ## Output Layer
        G_M = torch.cat((G, M), 2) # (bs, T, 10d)
        G_M = self.output_layer_1(G_M).squeeze() # (bs, T)
        G_M = F.pad(G_M, pad=(0,self.maxlen-G_M.size(1),0,0), mode='constant', value=0) # (bs, self.maxlen)

        logits = F.relu(self.output_layer_2(G_M)) # (N, 1000)
        logits = self.output_layer_3(logits) # (N, output_size)
        #print("M: ", M.size(), " G_M: ", G_M.size(), " logits: ", logits.size())
        
        #return F.softmax(logits, dim=-1) 
        return logits
    
    
    def loss(self, y, y_target):
        
        #loss = nn.MSELoss()(y,y_target)
        loss = nn.CrossEntropyLoss()(y,y_target)
        
        return loss


    def update(self, lr):
                
        update = torch.optim.SGD( self.parameters(), lr=lr )
        
        return update
    
    def update_learning_rate(self, optimizer, lr):
   
        for param_group in optimizer.param_groups:
            param_group['lr'] = lr

        return optimizer

### Generate Embeddings
Here you can choose your initial embeddings. Use either `pretrained embeddings` (glove/fasttext) or `uniform random embeddings`. **Uncomment the one you want to use.**

In [ ]:
#Generating Embeddings

embedding_size = 300

# embedding can be initialized from glove or fastest pretrained embeddings or from random ones.

#embeddings = gen_embeddings(word_dict, embedding_size, in_file='./data/glove/glove.6B.300d.txt', init_from='glove')
#embeddings = gen_embeddings(word_dict, embedding_size, in_file='./data/fasttext/wiki.en.vec', init_from='fasttext')
embeddings = gen_embeddings(word_dict, embedding_size, in_file=None, init_from='random')

In [ ]:
#Vectorizing and minibatch creation

trn_x1, trn_x2, trn_l, trn_y, max_d, max_q = vectorize(train_exps, word_dict, entity_dict)
train_data = gen_examples(trn_x1, trn_x2, trn_l, trn_y, 8, max_d, max_q)


dev_x1, dev_x2, dev_l, dev_y, _, _ = vectorize(dev_exps, word_dict, entity_dict)
dev_data = gen_examples(dev_x1, dev_x2, dev_l, dev_y, 8, max_d, max_q)

In [ ]:
#Model parameters

lr = 0.1
decay_rate = 1.2
dev_loss_old = 1000000
epochs_no = 10
drp = 0.25

In [ ]:
#Building model


net = Bidirectional_Attention_Model(embeddings=embeddings, hidden_size=128, output_size=len(entity_dict), maxlen=max_d, drp_rate=drp)
if torch.cuda.is_available():
    net.cuda()
opt = net.update(lr)

### Train the Model

In [ ]:
# save results in a .txt file
time_stamp = datetime.datetime.now().strftime("%y-%m-%d--%H-%M-%S")
file_name = 'logs'+'/'+time_stamp + "bidr" + ".txt"
file = open(file_name,"w",1) 
file.write(time_stamp+'\n\n') 
mystr = "Bidr_fasttext_gru"
file.write(mystr+'\n\n')

start = time.time()

for epoch in range(epochs_no):
    
    start_epoch = time.time()
    
    tr_loss, tr_acc = train_one_epoch(net,opt,train_data)
    
    dev_loss, dev_acc = test_one_epoch(net,dev_data)
    
    
    # update learning rate 
    if dev_loss > 0.99* dev_loss_old:
        lr /= decay_rate
    opt = net.update_learning_rate(opt, lr)
    dev_loss_old = dev_loss
    
    ep_details = ( 'Epoch {EP} [epoch length:{epoch_time:.0f}s | time from start:{fromstart:.1f}h] \t'
                   'lr={LR:.2e}\t'
                   'loss={train_loss:.3f}/{test_loss:.3f}\t'
                   'acc:{train_acc:.3f}/{test_acc:.3f} '.format(
                    EP=epoch,
                    epoch_time=time.time()-start_epoch,
                    fromstart=(time.time()-start)/3600,          
                    LR= lr,  
                    train_loss=tr_loss, test_loss=dev_loss,
                    train_acc=tr_acc, test_acc=dev_acc ) )
    
    
    
    
    file.write(ep_details+'\n')
    print(ep_details)
    
file.close()

### Test the model

In [ ]:
fin_test = "./data/cnn/test.txt"
test_exps = load_data(fin_test, relabeling=True)

tst_x1, tst_x2, tst_l, tst_y, max_d, max_q = vectorize(test_exps, word_dict, entity_dict)
test_data = gen_examples(tst_x1, tst_x2, tst_l, tst_y, 128, max_d, max_q)

In [ ]:
tst_loss, tst_acc = test_one_epoch(net,test_data)
print("Test loss and accuracy = ",tst_loss,"\t",tst_acc)

## Baseline Model

Our plan was to incrementally improve the model. So we first started with a simple model. The model **does not consider attention**. This is our baseline model. The following figure demonstrates the model.

<img src="images/baseline.jpg"> 
Figure: Baseline Model


In [ ]:
class Basic_LSTM_model(nn.Module):

    def __init__(self, embeddings, hidden_size, output_size, num_layers=1, drp_rate=0.25):
        super(Basic_LSTM_model, self).__init__()
        
        self.vocab_size = embeddings.shape[0]
        self.embedding_size = embeddings.shape[1]
        self.hidden_size = hidden_size
        self.output_size = output_size
        self.num_layers = num_layers
        self.drp = drp_rate
        
        weight = torch.Tensor(embeddings)
        self.embeddings = nn.Embedding.from_pretrained(weight, freeze=False)
            
        self.context_lstm = nn.LSTM(embedding_size, hidden_size, num_layers, 
                          batch_first=True, bidirectional=True, dropout=self.drp)
        
        self.linear = nn.Linear(hidden_size*2, output_size)

        
    def forward(self, doc_x, ques_x, doc_seq_lengths, ques_seq_lengths):   
        '''
        doc_x: torch tensor. size - (batch_size, doc_seq_len)
        ques_x: torch tensor. size - (batch_size, ques_seq_len)
        doc_seq_lengths: 1d numpy array containing lengths of each document in doc_x
        ques_seq_lengths: 1d numpy array containing lengths of each question in ques_x
        
        '''
        
        def contextual_embedding(data, seq_lengths):
            
            # Sort by length (keep idx)
            seq_lengths, idx_sort = np.sort(seq_lengths)[::-1], np.argsort(-seq_lengths)
            idx_original = np.argsort(idx_sort)
            idx_sort = torch.from_numpy(idx_sort).type(dtypeLong)
            data = data.index_select(0, idx_sort)

            packed_input = pack_padded_sequence(data, seq_lengths, batch_first=True)
            packed_output, (hidden, c) = self.context_lstm(packed_input)
            output, _ = pad_packed_sequence(packed_output, batch_first=True)
            #print("out: ", output.size(), " hid: ", hidden.size())

            #Un-sort by length
            idx_original = torch.from_numpy(idx_original).type(dtypeLong)
            output = output.index_select(0, idx_original)
            hidden = hidden.index_select(1, idx_original)
            c = c.index_select(1, idx_original)
            
            return output, hidden, c

        doc_data = self.embeddings(doc_x) # doc_data shape: (batch_size, doc_seq_len, embedding_dim)
        ques_data = self.embeddings(ques_x) # ques_data shape: (batch_size, ques_seq_len, embedding_dim)
         
        ##For Documents/questions
        doc_output, doc_hidden, doc_c = contextual_embedding(doc_data, doc_seq_lengths)
        ques_output, ques_hidden, ques_c = contextual_embedding(ques_data, ques_seq_lengths)
        # output shape: (batch_size, seq_len, hidden_size * num_directions)
        # hidden shape: (num_layers * num_directions, batch_size, hidden_size)
        
        #Obtaining the final output of documents
        docs_fwd_h = doc_hidden[0:doc_hidden.size(0):2]
        docs_bwd_h = doc_hidden[1:doc_hidden.size(0):2]
        docs_hidden = torch.cat([docs_fwd_h, docs_bwd_h], dim=2) 
        #docs_hidden shape:  torch.Size([1, bs, 256])
        
        #Obtaining the final output of questions
        ques_fwd_h = ques_hidden[0:ques_hidden.size(0):2]
        ques_bwd_h = ques_hidden[1:ques_hidden.size(0):2]
        ques_hidden = torch.cat([ques_fwd_h, ques_bwd_h], dim=2) 
        #ques_hidden shape:  torch.Size([1, bs, 256])
        
        final_op = docs_hidden.squeeze() + ques_hidden.squeeze() 
        #final_op shape:  torch.Size([bs, 256])
        
        logits = self.linear(final_op) #logits shape:  torch.Size([bs, numClasses])
        #print("logits shape: ", logits.size())
        
        return logits
    
    def loss(self, y, y_target):
        
        loss = nn.CrossEntropyLoss()(y,y_target)
        
        return loss


    def update(self, lr):
                
        update = torch.optim.SGD( self.parameters(), lr=lr )
        
        return update
    
    def update_learning_rate(self, optimizer, lr):
   
        for param_group in optimizer.param_groups:
            param_group['lr'] = lr

        return optimizer

### Generate Embeddings
Here you can choose your initial embeddings. Use either `pretrained embeddings` (glove/fasttext) or `uniform random embeddings`. **Uncomment the one you want to use.**

In [ ]:
#Generating Embeddings

embedding_size = 300

# embedding can be initialized from glove or fastest pretrained embeddings or from random ones.

#embeddings = gen_embeddings(word_dict, embedding_size, in_file='./data/glove/glove.6B.300d.txt', init_from='glove')
#embeddings = gen_embeddings(word_dict, embedding_size, in_file='./data/fasttext/wiki.en.vec', init_from='fasttext')
embeddings = gen_embeddings(word_dict, embedding_size, in_file=None, init_from='random')

In [ ]:
#Vectorizing and minibatch creation

trn_x1, trn_x2, trn_l, trn_y, max_d, max_q = vectorize(train_exps, word_dict, entity_dict)
train_data = gen_examples(trn_x1, trn_x2, trn_l, trn_y, 8, max_d, max_q)


dev_x1, dev_x2, dev_l, dev_y, _, _ = vectorize(dev_exps, word_dict, entity_dict)
dev_data = gen_examples(dev_x1, dev_x2, dev_l, dev_y, 8, max_d, max_q)

In [ ]:
#Model parameters

lr = 0.1
decay_rate = 1.2
dev_loss_old = 1000000
epochs_no = 10
drp = 0.25

In [ ]:
#Building model


net = Basic_LSTM_model(embeddings=embeddings, hidden_size=128, output_size=len(entity_dict),drp_rate=drp)
if torch.cuda.is_available():
    net.cuda()
opt = net.update(lr)

### Train the Model

In [ ]:
# save results in a .txt file
time_stamp = datetime.datetime.now().strftime("%y-%m-%d--%H-%M-%S")
file_name = 'logs'+'/'+time_stamp + "bidr" + ".txt"
file = open(file_name,"w",1) 
file.write(time_stamp+'\n\n') 
mystr = "Basic_LSTM"
file.write(mystr+'\n\n')

start = time.time()

for epoch in range(epochs_no):
    
    start_epoch = time.time()
    
    tr_loss, tr_acc = train_one_epoch(net,opt,train_data)
    
    dev_loss, dev_acc = test_one_epoch(net,dev_data)
    
    
    # update learning rate 
    if dev_loss > 0.99* dev_loss_old:
        lr /= decay_rate
    opt = net.update_learning_rate(opt, lr)
    dev_loss_old = dev_loss
    
    ep_details = ( 'Epoch {EP} [epoch length:{epoch_time:.0f}s | time from start:{fromstart:.1f}h] \t'
                   'lr={LR:.2e}\t'
                   'loss={train_loss:.3f}/{test_loss:.3f}\t'
                   'acc:{train_acc:.3f}/{test_acc:.3f} '.format(
                    EP=epoch,
                    epoch_time=time.time()-start_epoch,
                    fromstart=(time.time()-start)/3600,          
                    LR= lr,  
                    train_loss=tr_loss, test_loss=dev_loss,
                    train_acc=tr_acc, test_acc=dev_acc ) )
    
    
    
    
    file.write(ep_details+'\n')
    print(ep_details)
    
file.close()

### Test the model

In [ ]:
fin_test = "./data/cnn/test.txt"
test_exps = load_data(fin_test, relabeling=True)

tst_x1, tst_x2, tst_l, tst_y, max_d, max_q = vectorize(test_exps, word_dict, entity_dict)
test_data = gen_examples(tst_x1, tst_x2, tst_l, tst_y, 128, max_d, max_q)

In [ ]:
tst_loss, tst_acc = test_one_epoch(net,test_data)
print("Test loss and accuracy = ",tst_loss,"\t",tst_acc)

## Analysis of the Attention Based Models


### Numbers of parameters required for each of the models:


| Model                           | #Parameters                    | 
| ------------------------------- |:------------------------------:| 
| Baseline                        |  $36106816$ ($36.10$ million)| 
| Unidirectional Attention        |  $36061928$ ($36.06$ million)| 
| Bidirectional Attention         |  $39574738$ ($39.57$ million)| 

In [ ]:
def display_num_param(net):
    nb_param = 0
    for param in net.parameters():
        nb_param += param.numel()
    print('There are {} ({:.2f} million) parameters in this neural network'.format(nb_param, nb_param/1e6))
    print(nb_param)

display_num_param(net)

### Time Required for a single epoch 

Time required for a single epoch of our attention based models in `Nvidia GTX 1080 Ti`:

| Model                    | RNN Type | Single Epoch Time |
|-------------------------:|----------|-------------------|
| Baseline                 | GRU      | 0.3hr             |
| Unidirectional Attention | GRU      | 0.5hr             |
| Unidirectional Attention | LSTM     | 0.7hr             |
| Bidirectional Attention  | GRU      | 5.2hrs            |

## Results of the Attention Based Models

#### Baseline Model:

| RNN Type | Pretrained Embedding | Accuracy |
|----------|----------------------|----------|
| GRU      | Random               | $46.2$   |


#### Unidirectional Attention Model

| RNN Type | Pretrained Embedding | Accuracy |
|----------|----------------------|----------|
| GRU      | Random               | $66.87$  |
| GRU      | Glove                | $71.21$  |
| **GRU**  | **Fasttext**         | $72.29$  |
| LSTM     | Random               | $67.03$  |
| LSTM     | Glove                | $71.30$  |
| LSTM     | Fasttext             | $71.58$  |


#### Bidirectional Attention Model

| RNN Type | Pretrained Embedding | Accuracy |
|----------|----------------------|----------|
| GRU      | Random               | $55.69$  |
| GRU      | Glove                | $59.32$  |
| GRU      | Fasttext             | $60.09$  |


#### Best Result Variance

Our best result was obtained from `Unidirectional Attention Model` using `GRU` with `Fasttext pretrained embeddings`. The following table shows the variance of the accuracies for 3 runs of the best model.

|       | Accuracy |
|-------|----------|
| Run#1 | $71.57$    |
| Run#2 | $72.29$    |
| Run#3 | $71.69$    |
| **Mean**| **71.85**  |
| **Std**   | **0.39** |


## Result Analysis of the Attention Based Models


From the above tables, we can see that we get the best accuracy from `Unidirectional Attention Model` using `GRU` with `Fasttext pretrained embedding`. 

Our baseline model does not perform on a par with the attention models. The reason is obvious - the model is very simple. It is not possible to capture all the passage information in a single `Passage Vector`. So it performs badly compared to other models.

Our `Unidirectional Attention Model` performs quite good. We can see the efficacy of using `pretrained embeddings` like glove or fasttext over `random embeddings` from the results. `Fasttext` is outperforming `glove`. The reason behind this is - alongside contexts fasttext also considers `character n-grams`. This enables fasttext to generate better word embeddings for rare words. 
We tried with both GRU and LSTM. From the result, we can see there performances are almost similar. But LSTM takes more time to train.

Our `Bidirectional Attention Model` seemed to be promising. But it fails to produce a better result compared to `Unidirectional Attention Model`. We analyzed the outputs of the model and found that our `Bidirectional Attention Model` overfits the training data. That is why it fails to generalize and performs poorly on test data.


# Transformer Network Based Model

## Attention Over Attention Baseline Model

<img src="images/attention_over_attention_baseline.jpeg.jpg"> 

## In this baseline model, we do as follows:
1. **Contextual Embedding**:<br>
h(D)=bi-GRU(D)<br>
h(Q)=bi-GRU(Q)<br>
2. **Parwise Matching Score**: <br>
$M(i,j)=h_D(i)^Th_Q(j)$<br>
3. **Compute Query-to-Document Attention and vice-versa**:<br>
$\alpha(t)=softmax(M(1,t),M(2,t),...,M(|D|,t)$ for t=1,2,...|Q|<br>
$\beta(t)=softmax(M(t,1),M(t,2),...,M(t,|Q|)$ for t=1,2,...|D|<br>
for |D| is the number of tokens in document <br>
and |Q| is the number of tokens in query <br>
4. **Attention over Attention**:<br>
$\beta=\frac{1}{|D|}\sum_{t=1}^{|D|}\beta(t)$<br>
$s=\alpha^T\beta$<br>
5. **Prediction**: <br>
$P(w|D,Q)=\sum_{i \in I(w,D)}s_i$<br>
where I(w,D) indicate the position that word w appears in D

## Drawbacks of the Baseline Model:
-  As the number of tokens in the Document usually is very big, in this data, the max length is 2000, the gru may probably not capture the that long dependency well
- Not only that, gru may cause the running time very long, so we should find a way to make model run faster

    

## Proposed Model
-  Transformer is the model ultilizing the attention mechanism that compute the representation of sequence parallel:
<img src="images/Transformer_Model.jpg"> 

- As here we only need to **replace gru** by the different mechanism to overcome it cons, so we only need to use the Transformer Encoder part
<img src="images/Transformer_Encoder.jpg"> 

## Computatio of the Representation of Inputs in the Proposed Model:
-  Attention mechanism: $Attention(Q,K,V)=softmax(\frac{QK^T}{\sqrt{d_k}})V$ <br>
This one generalize the normal attention mechanism we know with RNN: for example, in the encoder-decoder scheme, K and V can be the representation of data in the encoder side, and Q is the vector we want to find the attention on K and V.
- Multi-head Attention: $Multihead(Q,K,V)=concat(head_0,head_1,head_h)W^O$ <br>
where $head_i=Attention(QW_i^Q,KW_i^K,VW_i^V)$

- Feed Forward: $FFN(x)=RELU(xW_1+b_1)W_2 +b_2$
- Positional Encoding: We represent position of the token by <br>
$PE(pos,2i)=sin(\frac{pos}{1000^\frac{2i}{d_{model}}})$<br>
$PE(pos,2i+1)=cos(\frac{pos}{1000^\frac{2i}{d_{model}}})$<br>
Please note that PE of pos here is a vector of size $d_{model}$ whose index 2i or 2i+1 is computed by those above formulas


Here, we only report the results because the proposed model has different data input pipeline and different program structure. Please refer to the separate notebook __[3_Transformer_Encoder_Attention_over_Attention_Notebook.ipynb]__ for more details.

Please find the Github repository for the Transformer Network Based Model [here](https://github.com/tungngthanh/Transformer_Encoder_AoAreader/blob/master/Analyze%20Model.ipynb).

## Results of the Proposed Model

|Randomized Embedding                          | Accuracy | Number of parameters | Training speed |   
|----------------------------------------------|----------|----------------------|----------------|
| Attention over Attention                     | 69.89%   | 46,835,392           | 13,874 s/epoch |   
| Transformer Encoder Attention over Attention | 70.86%   | 47,721,984           | 5,582 s/epoch  |   


**Result Analysis**
- In here we can see that Transformer Encoder gives a boost of around 1 percentage which show that Transformer Encoder is a good alternative to the GRU. 
- More over, one of the best properties of Transformer is allowing parallel computing as all of its components can do that. Here you can see that the training speed is faster, only around 1 and half hour for 1 epoch while the original took nearly 4 hours for 1 epoch.


# Pointer Network Based Model

## Motivation for the Proposed Model

Existing state-of-the-art solutions to the Cloze-form reading comprehension on the CNN dataset typically calculate a probability distribution over the entities in the vocabulary. Without using explicit information about entities (which basically provides an answer candidate list - unintended consequences of anonymising the entities), one way to solve the reading comprehension question-answering problem in an ideal case is to try to model the position/index of the answer in the input passage itself. Pointer Networks (Vinyals, 2015) is a solution that models attention directly over the input.

This is a pointer-networks inspired solution to RC, but there are various issues that come up with such an adaptation. One of the main problems is that the dataset provides the solution entity, but does not specify the exact context/sentence where this entity occurs - which means that the entity may appear any number of times in the passage, in different contexts, but we do not know which context is the most useful for predicting the answer to the query. This is problematic because it is necessary to know how many positions are being predicted during training.

## Model Description

The original pointer network attention is $v(\tanh (W_1 E + W_2 D))$ where E is the encoder output and D is the decoder output, and $W_1$, $W_2$ and V are outputs from linear layers. Here, the addition is changed to a dot product, since similarity between the passage (E) and query (D) representations (here obtained through a GRU layer) is required and SELU is used instead of tanh as the non-linear activation function. The final output is probabilities for each index in the passage obtained from a softmax operation.


Number of parameters: 12,717,312 (with POS-tags) 12,660,480 (without POS-tags)

## Training Strategies

Four different strategies were tried to find a loss that works in this scenario (multiple, differing number of ones in answer one-hot matrix), for which accuracies on the development set (without POS) are reported. The accuracy is calculated by getting the index with the highest predicted probability, and if it matches _any_ index in the list of known answer indices, it is considered a match - since the referred entity would be the same no matter which index in the list is predicted. 

(1) Train directly with one-hot answer matrix, and do 
    (a) similarity between passage and query final representations: Unsuccessful (dev acc: __0.4%__)
    (b) dot-product for each word in the passage with query representation. Unsuccessful, but works better with (4). (dev acc: __1.86%__)
    (c) similarity between passage and query as encoder output, with answer-matrix as decoder output. Unsuccessful. (dev acc: __1.04%__)

(2) Train with query expanded to passage length. Essentially trying to match query-length context in the passage; since the correct entity context is just one sentence in the passage. Unsuccessful, but works better when used with (4). (dev acc: __2.37%__)

(3) Train with one-hot answer matrix for first 3 epochs or so; then reduce the number of ones predicted in the answer by picking the top K predicted already to train. Reduce K for each epoch until K=1. Also tried with initialising net weights to identity matrices before training. The idea is, if the passage words and query similarity provides a strong enough signal, the answer entity with the correct context is more likely to be preferred early, and with repeated filtering, the answer entity in the correct context might be in the top K and eventually top 1. Unsuccessful. (dev acc: __1.03%__)

(4) Sum over the probabilities of all the known answer/gold indices. Target sum is one - basically predict zero probability for all words except answer entities, but individual probabilities are not important. Collectively try to maximise all index probabilities, under the assumption that the correct context will provide maximum contribution. Two parts to this: 
    (a) Train with decoder input=input_passage. Maximise sum(softmax(input_passage * attention weights)): Unsuccessful. (dev acc: __3.628%__) 
    (b) Remove decoder. Directly maximise sum(softmax(pointer_attention_weights)): somewhat (relatively) better. (dev acc: __6.67%__)

(5) __Possible future work__: 
    (a) Represent each word in passage by similarity with bi-directional query-size-window context, not including the word itself. Essentially, because the entities/placeholders are randomly initialised and have no similarity with the query, the dependence is entirely on the surrounding context. 
    (b) Model by stride: pick a reasonable stride size in the passage within which answer entity may not repeat. Learn attention weights for similarity with query with each stride then combine.


Here, we only report the proposed model description and the training strategies because the proposed model has different data input pipeline and different program structure. Please refer to the separate notebook __[4_Pointer_Network_Model_Notebook.ipynb]__ for more details.

## Analysis of the Result for Pointer Network Based Model

__Final test data accuracies__: __6.44%__ (without POS) __5.62%__ (with POS) 

Possible reasons for poor performance:
- Target is a unique, sparse matrix of document size, the relevant answer entities are randomly initialised, and to predict indices, the design does not necessarily provide signals from the most useful context.
- Loss calculation is not entirely straightforward.
- Using a better contextual embedding or more layers/parameters may be helpful. 


# Concluding Remarks

We tried different approaches to solve a Cloze form question answering problem, and both the best performing models perform at par with each other and use attention. Further improvement may be possible using better embedding techniques for words such as ELMo, Bert, etc. and by incorporating some linguistic information like part of speech tags or dependency information more effectively. 
